lib/sth_database.js

```diff
@@ -1841,6 +1841,490 @@ function getNotificationInfo(data, callback) {
+          servicePath: servicePath,
+          entityId: entityId,
+          entityType: entityType
+        }
+      ) :
+      getCollectionName4Events(
+        {
+          databaseName: databaseName,
+          servicePath: servicePath,
+          entityId: entityId,
+          entityType: entityType
+        }
+      );
+
+    database.collections(function (err, collections) {
+      if (err) {
+        if (callback) {
+          return process.nextTick(callback.bind(null, err));
+        }
+      }
+
+      var dropCollectionFunctions = [];
+      collections.forEach(function (collection) {
+        if (sthConfig.SHOULD_HASH) {
+          dropCollectionFunctions.push(async.apply(dropHashedCollection, collection.collectionName, data, options));
+        } else if (entityId) {
+          if (collection.collectionName === collectionName) {
+            dropCollectionFunctions.push(async.apply(dropCollection, collection.collectionName, service));
+          }
+        } else if (isAggregated) {
+          if (collection.collectionName.indexOf('.aggr') >= 0 &&
+            collection.collectionName.indexOf(collectionName.slice(0, collectionName.indexOf('undefined'))) >= 0) {
+            dropCollectionFunctions.push(async.apply(dropCollection, collection.collectionName, service));
+          }
+        } else {
+          if (collection.collectionName.indexOf('.aggr') < 0 &&
+            collection.collectionName.indexOf(collectionName.slice(0, collectionName.indexOf('undefined'))) >= 0) {
+            dropCollectionFunctions.push(async.apply(dropCollection, collection.collectionName, service));
+          }
+        }
+      });
+
+      async.parallel(
+        dropCollectionFunctions,
+        callback
+      );
+    });
+  }
+}
+
+/**
+ * Removes the data associated to the attributes specified in the provided data in case of collection per
+ *  attribute data model
+ * @param {object} data It is an object including the following properties:
+ *  - {string} service The service
+ *  - {string} servicePath The service path
+ *  - {string} entityId The entity id
+ *  - {string} entityType The entity type
+ *  - {string} attrName The attribute name. In case no attribute name is provided, all the attributes of the
+ *    provided entity are removed
+ * @param {object} options It is an object including the following properties:
+ *  - {boolean} isAggregated Flag indicating if the operation refers to aggregated data. Raw data otherwise
+ * @param callback The callback to call with error or the result of the operation
+ */
+function removeDataIfCollection4Attribute(data, options, callback) {
+  var service = data.service,
+      servicePath = data.servicePath,
+      entityId = data.entityId,
+      entityType = data.entityType,
+      attrName = data.attrName;
+  var isAggregated = options.isAggregated;
+
+  var databaseName = getDatabase(service);
+  var database = db.db(databaseName);
+  var collectionName = isAggregated ?
```

Comment:
This function may be a candidate also for the abstraction of the second branch I suggested above.